From 49cf102d1a076933766bb5fb48a8160d529c9353 Mon Sep 17 00:00:00 2001
From: Valery Volgutov <valery.volgutov@lge.com>
Date: Thu, 26 Jul 2018 13:01:20 +0900
Subject: [PATCH] Add guard to avoid recursively text layout call

setupTextLayout can be called twice or more times recursively.
When this function call and do q->setImplicitSize
function call. This function set implicit width and height
and after that emit some signals. After that we can to
enter in setupTextLayout again and full passed it.
After second setupTextLayout passed - QTextEngine clear self
data (lines). When we return to first setupTextLayout
we have incorrect data for text engine and crash occurs in

    QTextLine firstLine = visibleCount == 1 && elideLayout
            ? elideLayout->lineAt(0)
            : layout.lineAt(0);
    Q_ASSERT(firstLine.isValid());
    *baseline = firstLine.y() <--- HERE + firstLine.ascent();

because firstLine::y try to access to lines which cleared
previously.

Upstream-Status: Submitted [https://codereview.qt-project.org/#/c/235313/]
---
 src/quick/items/qquicktext.cpp   | 6 +++++-
 src/quick/items/qquicktext_p_p.h | 1 +
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/quick/items/qquicktext.cpp b/src/quick/items/qquicktext.cpp
index 97cc33b95f..2944c6b365 100644
--- a/src/quick/items/qquicktext.cpp
+++ b/src/quick/items/qquicktext.cpp
@@ -89,6 +89,7 @@ QQuickTextPrivate::QQuickTextPrivate()
     , layoutTextElided(false), textHasChanged(true), needToUpdateLayout(false), formatModifiesFontSize(false)
     , polishSize(false)
     , updateSizeRecursionGuard(false)
+    , layoutInProgress(false)
 {
     implicitAntialiasing = true;
 }
@@ -673,7 +674,7 @@ QRectF QQuickTextPrivate::setupTextLayout(qreal *const baseline)
             && q->widthValid()
             && (q->heightValid() || maximumLineCountValid);
 
-    if ((!requireImplicitSize || (implicitWidthValid && implicitHeightValid))
+    if (layoutInProgress || (!requireImplicitSize || (implicitWidthValid && implicitHeightValid))
             && ((singlelineElide && availableWidth() <= 0.)
                 || (multilineElide && q->heightValid() && availableHeight() <= 0.))) {
         // we are elided and we have a zero width or height
@@ -699,6 +700,7 @@ QRectF QQuickTextPrivate::setupTextLayout(qreal *const baseline)
         *baseline = fm.ascent();
         return QRectF(0, 0, 0, height);
     }
+    layoutInProgress = true;
 
     bool shouldUseDesignMetrics = renderType != QQuickText::NativeRendering;
     if (extra.isAllocated())
@@ -1156,6 +1158,8 @@ QRectF QQuickTextPrivate::setupTextLayout(qreal *const baseline)
     if (truncated != wasTruncated)
         emit q->truncatedChanged();
 
+    layoutInProgress = false;
+
     return br;
 }
 
diff --git a/src/quick/items/qquicktext_p_p.h b/src/quick/items/qquicktext_p_p.h
index efa45e0958..72b84c9911 100644
--- a/src/quick/items/qquicktext_p_p.h
+++ b/src/quick/items/qquicktext_p_p.h
@@ -176,6 +176,7 @@ public:
     bool formatModifiesFontSize:1;
     bool polishSize:1; // Workaround for problem with polish called after updateSize (QTBUG-42636)
     bool updateSizeRecursionGuard:1;
+    bool layoutInProgress:1; // Do not do layout again if previous call is not completed
 
     static const QChar elideChar;
 
